### PR TITLE
Add PrintChatMessageEvent, posted for *all* messages added to the chat

### LINKED ISSUE #2810: Handle ALL chat (local & remote)
I am working on a mod that streams the chat to a log file. I am currently using this listener class:

```
public class ChatListener {
        @SubscribeEvent
        public void onChatMessage(ClientChatReceivedEvent event) {
            String text = event.message.getUnformattedText();           
            appendLog(text);
        }
    }
```

I have noticed it doesn't completely handle chat.

For example command autocomplete, and chat lines added by other mods (in this case the same mod) with `Minecraft.getMinecraft().thePlayer.addChatComponentMessage(new ChatComponentText(text));` are ignored.

I would like to log these as well (at least chat lines added that way; i could do fine without autocomplete). I don't care about server side. Basically, I would log every line of text printed in the `[Client thread/INFO]: [CHAT]` stream.

Is there some event I am missing that handles all the chat shown to player?

### PR DESCRIPTION
**Pull request by blay09**
This adds an event that is called for all messages printed to the chat, so both server-side chat messages (which already pass through ClientChatReceivedEvent) as well as all other messages (including client-side ones, like the screenshot taken message).

Tested using a [simple test mod](http://pastebin.com/MvqgTRzs) that adds a fake timestamp and blocks messages containing "testus".

[More complex real world example](https://github.com/blay09/BetterMinecraftChat/blob/master/src/main/java/net/blay09/mods/bmc/handler/ChatHandler.java#L51-L104) to add custom data to chat messages, add timestamps, reformat messages etc.

Fixes #2810

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/gui/GuiNewChat.java.patch [hunk 1]
```diff
@@ -0,0 +1,13 @@
+--- ../src-base/minecraft/net/minecraft/client/gui/GuiNewChat.java
++++ ../src-work/minecraft/net/minecraft/client/gui/GuiNewChat.java
+@@ -134,6 +134,10 @@
+ 
+     public void func_146234_a(ITextComponent p_146234_1_, int p_146234_2_)
+     {
++        net.minecraftforge.client.event.PrintChatMessageEvent event = net.minecraftforge.client.ForgeHooksClient.onPrintChatMessage(p_146234_1_, p_146234_2_);
++        if(event.isCanceled()) return;
++        p_146234_1_ = event.getMessage();
```
Comment (killjoy1221) [inline patches/minecraft/net/minecraft/client/gui/GuiNewChat.java.patch]: You should null check this.

Comment (blay09) [inline patches/minecraft/net/minecraft/client/gui/GuiNewChat.java.patch]: It's documented as "may not be null" but I guess it won't hurt?

Comment (killjoy1221) [inline patches/minecraft/net/minecraft/client/gui/GuiNewChat.java.patch]: A `@Nonnull` would've been useful.

File: src/main/java/net/minecraftforge/client/event/PrintChatMessageEvent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import net.minecraft.util.text.ITextComponent;
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraftforge.fml.common.eventhandler.Event;
+
+/**
+ * This event is fired whenever a message is printed to the chat in {@link net.minecraft.client.gui.GuiNewChat#printChatMessageWithOptionalDeletion(ITextComponent, int)}
+ * This event is fired on the {@link net.minecraftforge.common.MinecraftForge#EVENT_BUS}
+ * This event is {@link Cancelable}
+ *
+ * {@link #message} contains the message that is printed to the chat (may not be null)
```
Comment (killjoy1221) [inline src/main/java/net/minecraftforge/client/event/PrintChatMessageEvent.java]: You should document this on the method, not the class.

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java [hunk 3]
```diff
@@ -65,22 +65,13 @@
...
-import net.minecraftforge.client.event.FOVUpdateEvent;
-import net.minecraftforge.client.event.GuiScreenEvent;
-import net.minecraftforge.client.event.ModelBakeEvent;
-import net.minecraftforge.client.event.MouseEvent;
-import net.minecraftforge.client.event.RenderGameOverlayEvent;
-import net.minecraftforge.client.event.RenderHandEvent;
-import net.minecraftforge.client.event.RenderWorldLastEvent;
-import net.minecraftforge.client.event.ScreenshotEvent;
-import net.minecraftforge.client.event.TextureStitchEvent;
+import net.minecraftforge.client.event.*;
```
Comment (kashike) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: ...

**Comment by beppe9000** [comment]
Great work!

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Pull request closed without merging**